doc/source/settings.rst
.. _settings:

The ANOVA analysis in details
=================================

.. contents::

About the settings
----------------------

When using the :class:`~gdsctools.anova.ANOVA` instance or you 
when using the :class:`~gdsctools.anova.ANOVAReport` to create an
HTML report (see :ref:`html`), all tunable settings are accessible from an
attribute called :attr:`settings`::

    from gdsctools import ANOVA, ic50_test
    gdsc = ANOVA(ic50_test)
    gdsc.settings

This :attr:`settings` attribute is an instance of :class:`gdsctools.settings.ANOVASettings`, which is fully documented in the reference. 


In previous sections, we have already seen a few of the settings. For example in the :ref:`HTML` section, we changed the default **directory** where HTML pages are saved to a user-defined value. 

It is also important to note that when calling ANOVAReport, the first argument
is an ANOVA instance that contains the settings. So, ANOVAReport will use that
settings automatically (may still be changed later). Consider this example::

    >>> from gdsctools import ANOVA, ic50_test, ANOVAReport
    >>> gdsc = ANOVA(ic50_test)
    >>> gdsc.settings.FDR_threshold = 15
    >>> results = gdsc.anova_all()

    >>> ar = ANOVAReport(gdsc, results)
    >>> ar.settings.FDR_threshold 
    15

Let us now introduce some of the most important settings.

.. _regression:

Regression analysis
-----------------------


By default, the regression uses a :term:`OLS` method. 4 Factors may be
taken into account depending on the content of the 
:class:`~gdsctools.readers.GenomicFeature` data set.

Here, we will use R syntax for simplicity. Depending on the data and the
:class:`gdsctools.anova.settings` one of the following formula will be used:


.. math:: Y \sim C(tissue) + C(media) + C(MSI) + Feature

.. math:: Y \sim C(tissue) + C(MSI) + Feature

.. math:: Y \sim C(MSI) + Feature

.. math:: Y \sim Feature




The default regression method is the :term:`OLS` method. It is also the
recommended method::

    settings.regression.method = 'OLS'

Future version will include other regression methods such as  Elastic Net, Ridge or Lasso::

    settings.regression.method = 'ElasticNet'
    settings.regression.method = 'Ridge'
    settings.regression.method = 'Lasso'

.. note:: Here the ElasticNet/Ridge/Lasso regression like the OLS one is made 
    for one drug and one feature (ODOF). The module :mod:`gdsctools.elastic_net` 
    uses a different set of input data using all features for one drug.

If you use anything else than OLS, you should then consider settings 1 or 2
additional settings::

    settings.regression.alpha
    settings.regression.L1_wt

See :class:`~gdsctools.anova.ANOVASettings` for details.

The regression analysis uses at most 4 factors: :term:`MSI`, Tissue, 
:term:`MEDIA` and Feature.  The latter is always included but others can be
tuned.

MSI factor
~~~~~~~~~~~~
If included in the genomic feature data set, MSI are included by default. 
However, you may exclude it by setting its value to False::

    settings.include_MSI_factor

If **MSI_FACTOR** column is not found in the Genomic Feature data set, the MSI factor will be excluded automatically and the parameter above set to False. 

.. warning:: If you force the MSI factor to True wherease there 
    is not enough data in the binary sets of the MSI factor, error
    will be raised.


MEDIA factor
~~~~~~~~~~~~~

If included in the genomic feature data set, MEDIA are included by default.
However, you may exclude it by setting its value to False::

    settings.include_MEDIA_factor

If **MEDIA_FACTOR** column is not found in the Genomic Feature data set, 
the MEDIA factor will be set automatically to False. 

Tissue factor
~~~~~~~~~~~~~~~~~

Another factor used in the regression (tissue) will
be automatically excluded if there is only one tissue (or none). If several
tissues are available, you can still exclude it from the regression analysis
by settings this parameter to anything different from the default value (PANCAN)::

    settings.analysis_type = PANCAN


Filtering
-----------

When performing the analysis for a given drug and feature, the regression may
not be performed if there is not enough statistics.

These parameters will influence the number of tests being performed (number of associations of drug vs feature in :meth:`~gdsctools.anova.ANOVA.anova_all`)::

    - minimum_nonna_ic50
    - MSI_feature_threshold
    - feature_factor_threshold
 
The first parameter indicates the minimum number of valid IC50 required for a given drug to be analysed. The current default value is 6.

The second parameter indicates the minimum size of the positive and negative
population when IC50 are filtered by MSI factor (defaults to 2). 

The third parameter indicates the minimum size of the positive and negative
population when IC50 are filtered by Feature factor (defaults to 3). 

This table summarizes the effect of these parameters:


.. image:: _static/regression_filtering.png

The left hand side table mimics the IC50 data. The first column should and last
3 rows are not to be included in an IC50 matrix (see :ref:`Data`) but are added
here as annotations for the following discussions.

When the regression analysis is performed for a given drug and a given features,
3 filtering are performed using the 3 parameters aforementionned. First, 
a minimum number of values are required (minimum_nonna_ic50 setting). The second drug (D_NA) is therefore not analysed. The second check is performed with respect to the MSI values. A drug can be analysed only if (once NA are discarded) the number of IC50s corresponding to positive and negative MSIs is greater or equal to **MSI_feature_threshold**.
In our example, the drugs in column **D_pMSI=0** and **D_pMSI=1** are therefore
discarded since they have zero and only one positive MSI, respectively. 

Finally, similarly to the MSI check, a drug/feature association is analysed if
the number of IC50s corresponding to positive and negative feature is or equal
to **feature_factor_threshold**.


Multiple testing corrections
------------------------------

By default, the multiple testing correction  is based on 
Benjamini–Hochberg (BH) method and can be set to other methods using ::

    settings.pval_correction_method

.. seealso:: :class:`~gdsctools.stats.MultipleTesting` for details.

The multiple testing is performed globally across all drugs and all cell
lines.This parameter is stored in ::

    settings.pvalue_correction_level

By default it is set to *global*. Set it to *local* to keep the multiple
correction at the drug level (ODAF).    

When you perform an ANOVA analysis, the multiple correction method is used to 
populate the results column named ANOVA_FEATURE_FDR. 

If you change your mind and wish to run the analysis with another method, 
you do not need to re-run the entire analysis. Instead, simply change the
method's name and call :meth:`anova_all` again. Only the multiple testing computation is
performed, skipping ANOVA testing, which have already been done.

::

    results = an.anova_all()
    an.settings.pvalue_correction_method = 'qvalue'
    results = an.anova_all()




.. index:: volcano

volcano plots
-----------------

The volcano plots are one of the main results of the analysis and summarizes
visually the significance of the different associations. 

It is part of the :class:`~gdsctools.anova_report.AnovaResults` class and is 
returned either by an ODAF or ADAF analysis:

.. plot::
    :include-source:
    :width: 80% 

    from gdsctools import ANOVA, ic50_test
    gdsc = ANOVA(ic50_test)
    res = gdsc.anova_all()
    res.volcano()


Here are some
parameters used to tune the plots and selection of significant events:

- **pvalue_threshold** is used to select significant hits. See :class:`~gdsctools.anova_report.ANOVAReport`. 
- **effect_threshold** is used to select significant hits as well.
- **FDR_threshold**   is used in :class:`gdsctools.volcano.VolcanoANOVA`
  (horizontal lines)
- **volcano_FDR_interpolation** uses interpolation to plot the FDR lines in the
  volcano plot.
- **volcano_additional_FDR_lines** : [0.01, 0.1, 10]

.. seealso:: :class:`~gdsctools.volcano.VolcanoANOVA`.

.. todo:: explain the effect of interpolation, link to the signed effects in the quickstart section

others
----------

See :class:`~gdsctools.settings.ANOVASettings` for the full listing.


.. note:: Some settings will be set automatically when calling some functions.
    For instance, if you call :meth:`anova.ANOVA.set_cancer_type` to a single
    tissue, then the analysis_type will be set to the tissue's name. If there 
    are not enough positive or negative MSI, the MSI factor will ignored.


gdsctools/settings.py
# coding=utf-8
# -*- python -*-
#
#  This file is part of GDSCTools software
#
#  Copyright (c) 2015 - Wellcome Trust Sanger Institute
#  All rights reserved
#
#  File author(s): Thomas Cokelaer <cokelaer@gmail.com>
#
#  Distributed under the BSD 3-Clause License.
#  See accompanying file LICENSE.txt distributed with this software
#
#  website: http://github.com/CancerRxGene/gdsctools
#
##############################################################################
"""Code related to the ANOVA analysis to find associations between drug IC50s
and genomic features"""
import pandas as pd
import numpy as np
from easydev import AttrDict
import easydev
from gdsctools import version


__all__ = ['ANOVASettings']

class ANOVASettings(AttrDict):
    """All settings used in :class:`gdsctools.anova.ANOVA` analysis

    This class behaves as a dictionary but values for a given
    key (setting) can be accessed and changed easily like an
    attribute:

    ::

        >>> from gdsctools import ANOVASettings
        >>> s = ANOVASettings()
        >>> s.FDR_threshold
        25
        >>> s.FDR_threshold = 20

    It is the responsability of the users or developers to check the validity
    of the settings by calling the :meth:`check` method. Note, however, that
    this method does not perform exhaustive checks.

    Finally, the method :meth:`to_html` creates an HTML table that can
    be added to an HTML report.

    .. note:: **for developers** a key can be changed or accessed to as if
       it was an attribute. This prevents some functionalities (such as copy()
       or property) to be used normaly hence the creation of the
       :meth:`check` method to check validity of the values rather than
       using properties.

    Here are the current values available.


    ========================= ================ ========================================
    Name                      Default          Description
    ========================= ================ ========================================
    include_MSI_factor        True             Include MSI in the regression
    feature_factor_threshold  3                Discard association where a
                                               genomic feature has less than 3
                                               positives or 3 negatives values
                                               (e.g., 0, 1 or 2)
    MSI_factor_threshold      2                Discard association where a MSI
                                               count has less than 2 positives
                                               or 2 negatives values (e.g., 0,
                                               or 1).
    analysis_type             PANCAN           Type of analysis. PANCAN means
                                               use all data. Otherwise, you must
                                               provide a valid tissue name to
                                               be found in the Genomic Feature
                                               data set.
    pvalue_correction_method    fdr            Type of p-values correction
                                               method used. Could be *fdr*,
                                               *qvalue*  or one accepted
                                               by
                                               :class:`~gdsctools.stats.MultipleTesting`

    pvalue_correction_level   global           Apply pvalue correction 
                                               globally. Can also be set to 
                                               'drug_level' to apply 
                                               corrections at drug level
                                               only.
    equal_var_ttest           True             Assume equal variance in the
                                               t-test
    minimum_nonna_ic50        6                Minimum number of IC50 required
                                               to perform an analysis for a
                                               given drug.
    fontsize                  25               Used in some plots for labels
    FDR_threshold             25               FDR threshold used in volcano
                                               plot and significant hits
    pvalue_threshold          0.001            Used to select significant hits see
                                               :class:`~gdsctools.anova.ANOVAReport`
    directory                 html_gdsc_anova  Directory where images and HTML
                                               documents are saved.
    savefig                   False            Save the figure or not (PNG format)
    effect_threshold          0                Used in the volcano plot. See
                                               :class:`~gdsctools.volcano.VolcanoPlot`
    ========================= ================ ========================================


    There are parameters dedicated to the regression method:

    ======================= ========= =========================================
    Name                    Default   Description
    ======================= ========= =========================================
    regression_method       OLS       Regression method amongst OLS, Ridge
                                      Lasso, ElasticiNet.
    regression_alpha        0.01      Fraction of penalty included. If 0,
                                      equivalent to OLS.
    regression_L1_wt        0.5       Fraction of the penalty given to L1
                                      penalty term. Must be between 0 and 1.
                                      If 0, equivalent to Ridge. If 1
                                      equivalent to Lasso
    ======================= ========= =========================================

    .. seealso:: :ref:`settings_filtering` or
        gdsctools.readthedocs.org/en/master/settings.html#filtering

        decrease the number of significant hits.

    """
    def __init__(self, **kargs):
        super(ANOVASettings, self).__init__(**kargs)

        ## ANALYSIS ---------------------------------
        # include MSI as a co-factor
        self.include_MSI_factor = True
        # number of positive samples required to perform the test
        self.feature_factor_threshold = 3
        # How many MSI samples must be present to perform the test
        self.MSI_factor_threshold = 2
        self.include_media_factor = False

        self.analysis_type = 'PANCAN'
        self.pvalue_correction_method = 'fdr'   # or qvalue
        self.pvalue_correction_level = 'global'   # or qvalue
        self.equal_var_ttest = True
        self.minimum_nonna_ic50 = 6

        # Visualisation and HTML related ---------------------
        self.fontsize = 25
        self.FDR_threshold = 25
        self.pvalue_threshold = 0.001
        self.directory = 'html_gdsc_anova'
        self.savefig = False
        self.effect_threshold = 0 # use in volcano
        self.volcano_additional_FDR_lines = [0.01, 0.1, 10]
        self.volcano_FDR_interpolation = True

        # ----------------------- regression related
        self.regression_method = 'OLS' # can be ElasticNet, LAsso, Ridge
        self.regression_alpha = 0.01
        self.regression_L1_wt = 0.5
        # uses statsmodels package
        # The fraction of the penalty given to the L1 penalty term. Must be
        # between 0 and 1 (inclusive). If 0, the fit is ridge regression. If
        # 1, the fit is the lasso.
        self.version = version
        self.animate = True

        for k, v in kargs.items():
            self[k] = v

    def check(self):
        """Checks the values of the parameters

        This may not be exhaustive. Right now, checks

         - MSI factor is boolean.
         - Regression.method in OLS/Ridge/Lasso/ElasticNet
         - FDR thresohld in [0,1]
         - pvalues_threshold in [0,inf[
         - effect_threshold in [0,inf[
         - pvalue_correction_method
         - etc
        """
        inrange = easydev.check_range
        inlist = easydev.check_param_in_list
        # check validity of the settings
        inlist(self.include_MSI_factor, [False, True], 'MSI')
        inrange(self.feature_factor_threshold, 0, np.inf)
        inrange(self.MSI_factor_threshold, 0, np.inf)

        # all those methods are from statsmodels.stats.multitest.multipletests
        inlist(self.pvalue_correction_method, ['bonferroni', 'sidak',
            'holm-sidak', 'simes-hochberg', 'hommel', 'fdr_bh',
            'fdr_tsbj', 'fdr_tskby', 'fdr'],
            'pvalue correction method')
        inlist(self.equal_var_ttest, [True, False], 'equal_var_ttest')
        inrange(self.minimum_nonna_ic50, 0, np.inf)
        inrange(self.FDR_threshold, 0, 100)
        inrange(self.pvalue_threshold, 0, np.inf)
        inrange(self.effect_threshold, 0, np.inf)

        # for now, if MSI is False, this cannot be a PANCAN analysis
        # but a cancer specific analysis
        if self.include_MSI_factor is False:
            assert self.analysis_type != 'PANCAN', \
                'If MSI factor is not included, the analysis must be cancer'+\
                ' specific (i.e., a tissue must be set.'

        valid_reg_meth = ['OLS', 'ElasticNet', 'Lasso', 'Ridge']
        inlist(self.regression_method, valid_reg_meth)

        inlist(self.pvalue_correction_level, ['global', 'drug_level'])

    def to_html(self):
        """Convert the sets of parameters into a nice HTML table"""
        data = self.copy()
        data['volcano_additional_FDR_lines'] = \
                str(data['volcano_additional_FDR_lines'])

        settings = pd.DataFrame(data, index=[0]).transpose()

        settings.reset_index(inplace=True)
        settings.columns = ['name', 'value']

        html = settings.to_html(header=True, index=False)
        return html

    def __str__(self):
        txt = ''
        for k, v in self.items():
            txt += '- %s: %s\n' % (k, v)
        return txt
